### PR TITLE
IGNITE-15236 Thin 3.0: Start client-handler module

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Start client-handler module when the node starts.
* Add integration test.

### CODE REVIEW

**Review by isapego — APPROVED** [review]
Looks good to me

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ITThinClientConnectionTest.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+    /**
+     * Gets the client handler module for the give node.
+     *
+     * @param nodeName Node name.
+     * @return Client handler module.
+     * @throws Exception When failed.
+     */
+    private static ClientHandlerModule getClientHandlerModule(String nodeName) throws Exception {
+        Field field = IgnitionImpl.class.getDeclaredField("nodesStartedComponents");
```
Comment (kgusakov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ITThinClientConnectionTest.java:158]: Why just didn't you set the client port in node config? As I see, other network ports are hardcoded in config already.

**Review by kgusakov — CHANGES_REQUESTED** [review]

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ITThinClientConnectionTest.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ITThinClientConnectionTest.java:158]: There is no guarantee that the server will bind to the given port.

I've actually used a hardcoded port initially and then spent some time debugging test hangs on TeamCity, where `10800` turned out to be occupied by something else.

Client handled has `port` + `portRange` and will bind to the first available port. The only way to ensure stable tests is to retrieve the actual bound port from the module.

File: modules/client/src/test/java/org/apache/ignite/client/AbstractClientTest.java [hunk 2]
```diff
@@ -46,7 +45,7 @@
 
     protected static ConfigurationRegistry configurationRegistry;
 
-    protected static ChannelFuture serverFuture;
+    protected static ClientHandlerModule serverModule;
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client/src/test/java/org/apache/ignite/client/AbstractClientTest.java:48]: Maybe I'm missing something (not much experience with JUnit), but `@BeforeAll` requires a static method: https://junit.org/junit5/docs/5.0.0/api/org/junit/jupiter/api/BeforeAll.html.

What is a better way to achieve a similar behavior - start the server once, run many client tests?

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientHandlerModule.java [hunk 3]
```diff
@@ -27,55 +28,76 @@
...
      *
-     * @param ignite Ignite.
-     * @param log Logger.
+     * @param igniteTables Ignite.
+     * @param registry Configuration registry.
      */
-    public ClientHandlerModule(Ignite ignite, Logger log) {
-        this.ignite = ignite;
-        this.log = log;
+    public ClientHandlerModule(
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientHandlerModule.java:62]: Fixed.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientHandlerModule.java [hunk 4]
```diff
@@ -27,55 +28,76 @@
...
-    public void prepareStart(ConfigurationRegistry sysCfg) {
-        registry = sysCfg;
+    /** {@inheritDoc} */
+    @Override public void start() {
+        if (channel != null)
+            throw new IgniteException("ClientHandlerModule is already started.");
+
+        try {
+            channel = startEndpoint().channel();
+        } catch (InterruptedException e) {
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientHandlerModule.java:80]: Fixed.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientHandlerModule.java [hunk 5]
```diff
@@ -27,55 +28,76 @@
...
-     * Starts the module.
+     * Returns the local address where this handler is bound to.
      *
-     * @return channel future.
-     * @throws InterruptedException If thread has been interrupted during the start.
+     * @return the local address of this module, or {@code null} if this module is not started.
      */
-    public ChannelFuture start() throws InterruptedException {
-        return startEndpoint();
+    public @Nullable SocketAddress localAddress() {
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientHandlerModule.java]: Fixed.

**Review by kgusakov — APPROVED** [review]

**Pull request merged**